rust/operator/src/hdfs_controller.rs

```diff
@@ -0,0 +1,678 @@
+                ..EnvVar::default()
+            },
+            EnvVar {
+                name: "POD_NAME".to_string(),
+                value_from: Some(EnvVarSource {
+                    field_ref: Some(ObjectFieldSelector {
+                        field_path: String::from("metadata.name"),
+                        ..ObjectFieldSelector::default()
+                    }),
+                    ..EnvVarSource::default()
+                }),
+                ..EnvVar::default()
+            },
+            EnvVar {
+                name: "ZOOKEEPER".to_string(),
+                value_from: Some(EnvVarSource {
+                    config_map_key_ref: Some(ConfigMapKeySelector {
+                        name: Some(hdfs.spec.zookeeper_config_map_name.clone()),
+                        key: "ZOOKEEPER".to_string(),
+                        ..ConfigMapKeySelector::default()
+                    }),
+                    ..EnvVarSource::default()
+                }),
+                ..EnvVar::default()
+            },
+        ]),
+        volume_mounts: Some(vec![
+            VolumeMount {
+                mount_path: "/data".to_string(),
+                name: "data".to_string(),
+                ..VolumeMount::default()
+            },
+            VolumeMount {
+                mount_path: "/stackable/hadoop/etc/hadoop".to_string(),
+                name: "config".to_string(),
+                ..VolumeMount::default()
+            },
+        ]),
+        ports: Some(
+            rolegroup_ports
+                .iter()
+                .map(|(name, value)| ContainerPort {
+                    name: Some(name.clone()),
+                    container_port: *value,
+                    protocol: Some("TCP".to_string()),
+                    ..ContainerPort::default()
+                })
+                .collect(),
+        ),
+        ..Container::default()
+    })
+}
+
+fn local_disk_claim(name: &str, size: Quantity) -> PersistentVolumeClaim {
+    PersistentVolumeClaim {
+        metadata: ObjectMeta {
+            name: Some(name.to_string()),
+            ..ObjectMeta::default()
+        },
+        spec: Some(PersistentVolumeClaimSpec {
+            access_modes: Some(vec!["ReadWriteOnce".to_string()]),
+            resources: Some(ResourceRequirements {
+                requests: Some(BTreeMap::from([("storage".to_string(), size)])),
+                ..ResourceRequirements::default()
+            }),
+            ..PersistentVolumeClaimSpec::default()
+        }),
+        ..PersistentVolumeClaim::default()
+    }
+}
+
+fn hadoop_config_xml<I: IntoIterator<Item = (K, V)>, K: AsRef<str>, V: AsRef<str>>(
+    kvs: I,
+) -> String {
+    use std::fmt::Write;
+    let mut xml = "<configuration>\n".to_string();
+    for (k, v) in kvs {
+        writeln!(
+            xml,
+            "<property><name>{}</name><value>{}</value></property>",
+            k.as_ref(),
+            v.as_ref()
+        )
+        .unwrap();
```

Comment:
use an error instead?

Reply:
Writing to a String is infallible so this is ok.